README.md
kylin-installer
===============

Attention
=========
It supports unity and mate for Ubuntu/Ubuntu Kylin.

Depends
=======
debhelper (>= 9),python3,python3-pyqt5,python3-apt,python3-gi,python3-pil,python3-xdg,python3-piston-mini-client

GitHub
=======
https://github.com/UbuntuKylin/kylin-installer

FT Version:
changelog like this:
kylin-installer (0.0.7kord) juniper; urgency=medium


debian/changelog
kylin-installer (0.2.0kord) v100; urgency=medium

  * 卸载软件时增加了软件是否正在使用的判断
  * 修复安装百度网盘后点击“立即体验”不生效的问题
  * 修复v10.1上4k不生效的问题
  * 修复安装brower360界面显示不全的问题
  * 修复包的描述信息不规范的问题

 -- lijiang (rain) <lijiang@kylinos.cn>  Sat, 05 Dec 2020 17:41:51 +0800

kylin-installer (0.1.9kord) v100; urgency=medium

  * 修改ukui3主题是否存在的判断方式
  * 修改kylin-settings配置文件，支持默认双击打开

 -- lijiang (rain) <lijiang@kylinos.cn>  Sat, 28 Nov 2020 14:54:44 +0800

kylin-installer (0.1.8kord) v100; urgency=medium

  * 修复version的翻译问题

 -- lijiang (rain) <lijiang@kylinos.cn>  Thu, 26 Nov 2020 11:29:45 +0800

kylin-installer (0.1.7kord) v100; urgency=medium

  * 修改上个版本命令行安装失败的问题

 -- lijiang (rain) <lijiang@kylinos.cn>  Sat, 21 Nov 2020 13:36:42 +0800

kylin-installer (0.1.6kord) v100; urgency=medium

  * 修改了包名、版本信息显示方式，可以动态调整长度

 -- lijiang (rain) <lijiang@kylinos.cn>  Fri, 20 Nov 2020 17:32:43 +0800

kylin-installer (0.1.5kord) v100; urgency=medium

  * 增加了应用安装完成后点击立即体验软件可以自动退出的机制

 -- lijiang (rain) <lijiang@kylinos.cn>  Tue, 17 Nov 2020 11:18:56 +0800

kylin-installer (0.1.4kord) v100; urgency=medium

  * 修复v10-u系统中没有屏蔽主题切换的问题
  * 解决了可以打开多个应用的问题
  * 安装卸载后界面调整

 -- lijiang (rain) <lijiang@kylinos.cn>  Tue, 27 Oct 2020 09:02:19 +0800

kylin-installer (0.1.3kord) v100; urgency=medium

  * 增加软件国际化，支持中英文。
  * 卸载完成后，界面调整
  * 解决任务栏中无图标的bug

 -- lijiang (rain) <lijiang@kylinos.cn>  Fri, 25 Sep 2020 17:33:31 +0800

kylin-installer (0.1.2kord) v100; urgency=medium

  * 新增dbus后台卸载功能，完成卸载ui界面。
  * 新增软件安装后“立即体验”的功能

 -- lijiang (rain) <lijiang@kylinos.cn>  Thu, 17 Sep 2020 16:52:15 +0800

kylin-installer (0.1.1kord) v100; urgency=medium

  * 修改ui界面，调整界面大小；
  * 新增主题切换功能，可软件自己切换也可以跟随主题修改；
  * 修改按钮样式跟随系统，增加高亮反白处理；
  * 增加ukui主题判断，v10系统增加额外处理。

 -- lijiang (rain) <lijiang@kylinos.cn>  Thu, 17 Sep 2020 16:52:15 +0800

kylin-installer (0.1.0kord) v100; urgency=medium

  * 修复禅道bug-18171,bug-16181,替换软件图标

 -- lijiang (rain) <lijiang@kylinos.cn>  Fri, 11 Sep 2020 09:09:05 +0800

kylin-installer (0.0.9kord) v101; urgency=medium

  * 修复禅道bug-17486,bug-17452

 -- lijiang (rain) <lijiang@kylinos.cn>  Wed, 09 Sep 2020 11:49:46 +0800

kylin-installer (0.0.8kord) v100; urgency=medium

  * 修复右键选择打开方式图标不显示的bug

 -- lijiang (rain) <lijiang@kylinos.cn>  Tue, 01 Sep 2020 16:56:28 +0800

kylin-installer (0.0.7kord) v100; urgency=medium

  * 修改无法安装包含中文字符包的bug

 -- lijiang (rain) <lijiang@kylinos.cn>  Tue, 01 Sep 2020 16:56:28 +0800

kylin-installer (0.0.6kord) v100; urgency=medium

  * 修改包名

 -- lijiang (rain) <lijiang@kylinos.cn>  Mon, 31 Aug 2020 20:00:07 +0800

kylin-installer (0.0.5kord) v100; urgency=medium

  * 新增dbus后台，支持安装、卸载
  * 新增label的tooltips
  * 新增单例检测

 -- lijiang (rain) <lijiang@kylinos.cn>  Mon, 31 Aug 2020 20:00:07 +0800

kylin-installer (0.0.4kord) v100; urgency=medium

  * 去除菜单栏显示

 -- dengnan <dengnan@kylinos.cn>  Fri, 21 Aug 2020 17:31:30 +0800

kylin-installer (0.0.3kord) v100; urgency=medium

  * 修改主文件权限 

 -- dengnan <dengnan@kylinos.cn>  Fri, 07 Aug 2020 21:38:00 +0800

kylin-installer (0.0.2kord) v100; urgency=medium

  * deb包安装软件创建初始化

 -- dengnan <dengnan@kylinos.cn>  Fri, 07 Aug 2020 19:13:08 +0800

kylin-installer (0.0.1kord) trusty; urgency=low

  * Initial release (LP: #1293299).

 -- Jun Ma <majun@ubuntukylin.com>  Thu, 20 Mar 2014 10:14:56 +0800


ui/mainwindow.py
# -*- coding: utf-8 -*-

# Form implementation generated from reading ui file 'confw.ui'
#
# Created by: PyQt5 UI code generator 4.11.4
#
# WARNING! All changes made in this file will be lost!
import math

from PyQt5.QtWidgets import *
from PyQt5.QtGui import *
from PyQt5.QtCore import *
from PyQt5 import QtCore, QtGui, QtWidgets
from PyQt5.QtWidgets import *
try:
    _fromUtf8 = QtCore.QString.fromUtf8
except AttributeError:
    def _fromUtf8(s):
        return s

try:
    _encoding = QApplication.UnicodeUTF8
    def _translate(context, text, disambig):
        return QApplication.translate(context, text, disambig, _encoding)
except AttributeError:
    def _translate(context, text, disambig):
        return QApplication.translate(context, text, disambig)


import gettext
gettext.bindtextdomain("kylin-installer",  "/usr/share/locale")
gettext.textdomain("kylin-installer")
_ = gettext.gettext

class Ui_MainWindow(QWidget):
    def setupUi(self, MainWindow):
        #取消button灰色背景
        self.palette = QPalette()
        self.ColorPlaceHolderText = QColor(255,255,255,0)
        self.brush = QBrush()
        self.brush.setColor(self.ColorPlaceHolderText)
        self.palette.setBrush(QPalette.Button, self.brush)
        self.palette.setBrush(QPalette.ButtonText, self.brush)

        MainWindow.setObjectName(_fromUtf8("MainWindow"))
        MainWindow.resize(610,390)
        MainWindow.setAttribute(QtCore.Qt.WA_TranslucentBackground)
        MainWindow.setWindowFlags(Qt.FramelessWindowHint)

        self.centralWidget=QWidget(MainWindow)
        self.centralWidget.setObjectName(_fromUtf8("centralWidget"))
        self.centralWidget.setGeometry(QtCore.QRect(10, 10, 590, 370))

        self.message = QLabel(self.centralWidget)
        self.message.setObjectName(_fromUtf8("message"))
        self.message.setGeometry(95, 98, 400, 80)
        # self.message.setStyleSheet("QLabel{background-color:yellow}")
        self.messageLayout = QHBoxLayout(self.message)
        self.buttonLayout = QHBoxLayout(self.centralWidget)

        self.title=QWidget(self.centralWidget)
        self.title.setObjectName(_fromUtf8("title"))
        self.title.setGeometry(QtCore.QRect(0, 0, 500, 36))
        # self.title.setStyleSheet(".QWidget{background-color:#fafafa;}")

        self.title_icon=QLabel(self.title)
        self.title_icon.setObjectName(_fromUtf8("title_icon"))
        self.title_icon.setGeometry(QtCore.QRect(15, 10, 24, 24))
        self.title_icon.setStyleSheet("QLabel{;background-image:url('res/logo.svg')}")

        self.title_text=QLabel(self.title)
        self.title_text.setObjectName(_fromUtf8("title_text"))
        self.title_text.setGeometry(QtCore.QRect(45,10,200,24))
        self.title_text.setText(_("kylin-application-installer"))
        self.title_text.setContentsMargins(0,0,0,0)
        self.title_text.setAlignment(Qt.AlignLeft | Qt.AlignTop)
        # self.title_text.setStyleSheet("QLabel{background-color:red}")
        self.title_text_ft = QFont()
        # self.title_text_ft.setBold(True)
        self.title_text_ft.setPixelSize(14)
        self.title_text_ft.setFamily("Microsoft YaHei")
        self.title_text.setFont(self.title_text_ft)
        # self.title_text.setAlignment(Qt.AlignLeft | Qt.AlignBottom)

        #设置安装的软件图标
        self.icon=QLabel(self.centralWidget)
        self.icon.setObjectName(_fromUtf8("icon"))
        self.icon.setFixedSize(64, 64)

        self.tips=QLabel()
        self.tips.setObjectName(_fromUtf8("tips"))
        self.tips.setFixedSize(200, 80)
        self.tips.setContentsMargins(0,0,0,0)
        # self.tips.setStyleSheet("QLabel{background-color:blue}")
        #软件的包名
        self.pkgname=QLabel(self.tips) #软件包名label
        self.pkgname.setObjectName(_fromUtf8("pkgname"))
        self.pkgname.setContentsMargins(0,0,0,0)

        self.pkgname_ft = QFont()
        self.pkgname_ft.setFamily("Microsoft YaHei")
        self.pkgname_ft.setPixelSize(26)
        self.pkgname.setFont(self.pkgname_ft)
        self.pkgname.setGeometry(QtCore.QRect(0, 8, 200, 30))
        self.pkgname.setAlignment(Qt.AlignLeft | Qt.AlignBottom)
        self.pathwidth = QFontMetrics(self.pkgname.font())

        # self.pkgname.setText("茄子摄像头")

        #设置软件版本信息
        self.version=QLabel(self.tips)
        self.version.setObjectName(_fromUtf8("version"))
        self.version.setGeometry(QtCore.QRect(0, 46, 250, 15))
        self.version.setContentsMargins(0,0,0,0)
        # self.version.setAlignment(Qt.AlignLeft | Qt.AlignBottom)
        self.version_ft = QFont()
        self.version_ft.setFamily("Microsoft YaHei")
        self.version_ft.setPixelSize(12)
        self.version.setFont(self.version_ft)
        # self.pkgname.setStyleSheet("QLabel{background-color:red}")
        # self.version.setStyleSheet("QLabel{background-color:blue}")
        self.tips_layout = QVBoxLayout()
        self.tips_layout.setContentsMargins(0,0,0,0)
        self.tips_layout.setSpacing(0)
        self.tips_layout.addStretch()
        self.tips_layout.addWidget(self.pkgname)
        self.tips_layout.addSpacing(5)
        self.tips_layout.addWidget(self.version)
        self.tips_layout.addStretch()
        self.tips.setLayout(self.tips_layout)

        self.messageLayout.addStretch()
        self.messageLayout.addWidget(self.icon)
        self.messageLayout.addSpacing(15)
        self.messageLayout.addWidget(self.tips)
        self.messageLayout.addStretch()
        self.messageLayout.setContentsMargins(0,0,0,0)
        self.messageLayout.setSpacing(0)
        self.message.setLayout(self.messageLayout)

        # 菜单栏
        self.menuButton = QToolButton(self.centralWidget)
        self.menuButton.setGeometry(487, 5, 30, 30)
        self.menuButton.setIconSize(QSize(16,16))
        self.menuButton.setObjectName(_fromUtf8("menuButton"))
        self.menuButton.setIcon(QIcon.fromTheme("open-menu-symbolic"))
        self.menuButton.setFocusPolicy(Qt.NoFocus)
        self.menuButton.setPopupMode(QToolButton.InstantPopup) #未设置该属性会出现下拉箭头
        self.menuButton.setPalette(self.palette)
        self.menuButton.setProperty("useIconHighlightEffect", True)
        self.menuButton.setProperty("iconHighlightEffectMode", 1)
        # self.menuButton.hide()

        # 最小化按钮
        self.minButton = QPushButton(self.centralWidget)
        self.minButton.setGeometry(QtCore.QRect(521, 5, 30, 30))
        self.minButton.setIconSize(QSize(14,14))
        self.minButton.setObjectName(_fromUtf8("minButton"))
        # self.minButton.setIcon(QIcon.fromTheme("window-minimize-symbolic"))
        self.minButton.setFocusPolicy(Qt.NoFocus)
        self.minButton.setPalette(self.palette)
        self.minButton.setProperty("useIconHighlightEffect", True)
        self.minButton.setProperty("iconHighlightEffectMode", 1)

        # 关闭软件按钮
        self.closeButton = QPushButton(self.centralWidget)
        self.closeButton.setGeometry(QtCore.QRect(555, 5, 30, 30))
        self.closeButton.setIconSize(QSize(14,14))
        self.closeButton.setObjectName(_fromUtf8("closeButton"))
        self.closeButton.setFocusPolicy(Qt.NoFocus)
        # self.closeButton.setIcon(QIcon.fromTheme("window-close-symbolic"))
        self.closeButton.setPalette(self.palette)
        self.closeButton.setProperty("useIconHighlightEffect", True)
        self.closeButton.setProperty("iconHighlightEffectMode", 1)
        # self.closeButton.setFlat(True)

        self.install=QPushButton(self.centralWidget)
        self.install.setObjectName(_fromUtf8("pkgname"))
        self.install.setGeometry(QtCore.QRect(220, 234, 150, 45))
        self.install.setStyleSheet("QPushButton{background-color:#2d8ae1;border:0px;font-size:16px;border-radius:4px;color:#ffffff}"
                                   "QPushButton:hover{background-color:#3580c4;border:0px;border-radius:4px;font-size:16px;color:#ffffff}")

        self.loding = QMovie("./res/loading1.gif")
        self.progressBar = QLabel(self.centralWidget)
        self.progressBar.setObjectName(_fromUtf8("closeButton"))
        self.progressBar.setGeometry(QtCore.QRect(120, 255, 350, 4))
        self.progressBar.setMovie(self.loding)
        # self.progressBar.setTextVisible(False)
        # self.progressBar.setVisible(False)
        self.progressBar.setStyleSheet("QProgressBar{background-color:#e5e5e5;}")

        self.status=QLabel(self.centralWidget)
        self.status.setObjectName(_fromUtf8("status"))
        self.status.setGeometry(QtCore.QRect(246, 274, 98, 24))
        self.status_ft=QFont()
        self.status_ft.setPixelSize(12)
        self.status_ft.setFamily("Microsoft YaHei")
        # self.status_ft.setBold(True)
        self.status.setFont(self.status_ft)
        # self.status.setStyleSheet(".QLabel{background-color:transparent;border:0px;font-size:12px;}")
        self.status.setAlignment(Qt.AlignLeft | Qt.AlignBottom)
        self.status.hide()

        # 安装结束后的提示图标
        self.status_icon = QLabel(self.centralWidget)
        self.status_icon.setObjectName(_fromUtf8("status_icon"))
        self.status_icon.setGeometry(QtCore.QRect(186, 101, 48, 48))
        self.status_icon.hide()

        # 安装结束后的提示信息
        self.status_text = QLabel(self.centralWidget)
        self.status_text.setObjectName(_fromUtf8("status_text"))
        self.status_text.setGeometry(249, 107, 200, 47)
        self.status_text.setContentsMargins(0,0,0,2)
        # self.status_text.setStyleSheet("QLabel{background-color:red}")
        self.status_text.setAlignment(Qt.AlignLeft | Qt.AlignBottom)
        self.status_text.hide()

        self.status_text_ft = QFont()
        self.status_text_ft.setPixelSize(36)
        self.status_text_ft.setFamily("Microsoft YaHei")
        # self.status_text_ft.setBold(True)
        self.status_text.setFont(self.status_text_ft)
        # self.status_ft.set
        # 菜单栏
        self.menu = Menu(self.centralWidget)
        self.menuButton.setMenu(self.menu)

class Menu(QMenu):
    def __init__(self,parent):
        super(Menu, self).__init__(parent)
        self.group = QActionGroup(self)
        self.group.setExclusive(True)
        self.dark_model_action = self.addAction(_('dark model'))
        self.dark_model_action.setCheckable(True)
        self.group.addAction(self.dark_model_action)
        self.white_model_action = self.addAction(_('white model'))
        self.white_model_action.setCheckable(True)
        self.group.addAction(self.white_model_action)
        # self.help_action = self.addAction(str('帮助'))
        # self.help_action.setCheckable(True)
        # self.group.addAction(self.help_action)
        # self.abort_action = self.addAction(str('关于'))
        # self.abort_action.setCheckable(True)
        # self.group.addAction(self.abort_action)

